whole-tale/girder_wholetale
# Add initial support for external accounts

This PR providers a basic infrastructure for supporting auth integrations with 3rd party services using either OAuth flow, or API Key model (and DataONE which is a combination of both).

### Implementation details
The majority of the functionality is exposed via a single REST API endpoint: `GET /account?redirect=`. It returns an unsorted list of supported 3rd party services that user can authorize or is authorized with, as indicated by `state` field. Depending on `type` of auth provider (which can be `apikey`, `bearer` and `dataone`) there is a different flow that client UI should follow, as described below

#### OAuth (`type == 'bearer'`)
1. Authorization in OAuth flow, boils down to allowing user to follow a specially crafted href that jumps to OAuth IdP and then goes back to point of origin. If (`state == 'unauthorized'`), `url` field contains that url.
2. Deauthorization can be performed via `GET /account/:provider/revoke`. 

#### API Key (`type == 'apikey'`)
1. Authorization requires user to provide a key-value pair, where value is the API key, and the key is the `resource_server` that the former will be used with. A list of supported resource servers can be obtained by calling `GET /account/:provider/targets` e.g. `GET /account/zenodo/target`. After the key-value pair is provided by the user, it should be stored using `POST /account/:provider/key?resource_server=...&key=...`
2. Deauthorization is the same as in OAuth case, with the addition of `resource_server` as a required qarg.

#### DataONE (`type == 'dataone'`)
Authorization flow is similar to OAuth initially: you follow the url that accepts a redirect to a callback. The trick is: it doesn't return any of the usual information information, i.e. code nor token that can be used to obtain `access_token`. Instead DataONE sets a cookie on their page with _JSESSIONID_ that can be used to call a separate endpoint (`/portal/token`) to grab a JWT. I don't see a programmatic way of obtaining _JSESSIONID_ and that's why it has to happen on the client side... I came up with the following workflow:
1. User follows the `url` to authenticate with DataONE using ORCID authorization.
2. `url` sets the usual `redirect` to `/account/:provider/callback`, and during callback a bogus token is created:
   ```
   {
       "provider": "dataone...",
       "resource_server": <cn_netloc>,
       "token_type": "dataone-pre",
       "access_token": "",
   }
   ```
3. Whenever a user token of type _"dataone-pre"_ is detected in `GET /account`, the provider state is set to _"preauthorized"_
4. On the UI side, the settings page performs a `XMLHttpRequest` to obtain a JWT for each provider in state _"preauthorized"_ (see https://github.com/whole-tale/girder_wholetale/blob/ext_keys/web_client/views/ExtKeysView.js#L95-L125).
5. If a valid JWT is obtained in 4., it's stored as if it was an API key using `POST 'account/dataone.../key`. This updates both `token_type` and `access_token` in the existing token.

#### Notes
* `targets` keyword on providers returned by `GET /account` is only populated for authorized api key providers (check it's value during step 14. from *How to test?*) 

#### Requirements

* Grab ORCID client id/secret from `#core-dev` and add the following to your `.bashrc`:
   ```
   export ORCID_CLIENT_ID="APP-..."
   export ORCID_CLIENT_SECRET="6..."
   ```
   Make sure you're running latest `deploy-dev`. Run `source ~/.bashrc` before `make dev`.

### How to test?

1. Deploy locally, log in to Girder UI using Globus OAuth.
2. Navigate to _External Keys_ view (top right menu under your username) https://girder.local.wholetale.org/#ext_keys Should look like this (2 buttons + Zenodo + Dataverse):
![extkeys](https://user-images.githubusercontent.com/352673/67789212-98823880-fa41-11e9-804c-2697ee2465a1.png)

3. Open a terminal and run (keep it handy for verification that things work):
   ```
   docker exec -ti $(docker ps --filter=name=wt_girder -q) girder-shell
   ```
4. In terminal run:
   ```
   from girder.models.user import User
   User().findOne({'login': 'kowalikk'})['otherTokens']   # NOTE: change the login value
   ```
   At this point it should contain only Globus Transfer token.
5. On _External Keys_ page click on ORCID button.
6. After completed flow button should be visible in the 2nd row and 
   ```
   User().findOne({'login': 'kowalikk'})['otherTokens']
   ```
   should contain additional token.
7. Click on ORCID button in 2nd row.
8. It should return to the first row, token should be gone from user model.
9. Repeat 5. - 8. for DataONE
10. Click on + right of "Zenodo". Modal should open. Available choices for *Repository URL* should be: "zenodo.org" and "sanbox.zenodo.org".
11. Select "sandbox.zenodo.org", provide any value as API KEY, click "Connect"
12. Verify that "sandbox.zenodo.org" is displayed under Zenodo and it has red "trash can" button. Verify that user model has the appropriate token (check `resource_server` in particular).
13. Click on + right of "Zenodo" again. Modal should open. Available choices for *Repository URL* should be: "zenodo.org".
14. Repeat 11. - 12. using "zenodo.org"
15. Click on the red "trash can" button and confirm that selected resource disappear from UI and user model.
